### ISSUE TITLE
idea

### ISSUE THREAD

**Issue by prographo** [philanc/plc#15]
PBKDF2 is SHA1 only, can it do SHA512 also?

**Comment by catwell** [comment]
It's pretty easy to implement. I can probably do it someday for fun if you need it :)

**Comment by philanc** [comment]
I already considered implementing a key derivation function (KDF). As catwell said above, it would be pretty easy, but I think that it would also be _useless_:

A KDF is a function that uses a lot of CPU (and lots of memory, for modern KDFs such as scrypt or argon2) to make it hard for example to brute-force a password.

Typically, a PBKDF2 call would use a number of iterations high enough so the computation takes for example 1 second, for a C implementation on an average X86_64 box.

Assuming that a pure Lua implementation is, let say 1,000 times slower than a C implementation, the PBKDF2 call with the same parameters would take approximately 20 minutes!  which is just not practical.

Reducing the number of iterations to something practical for the Lua implementation (say 1 sec) would make it computable in 1 msec on the same average PC (not to mention brute-forcing using GPUs, FPGAs or even ASICs...), which cannot be considered as reasonable in terms of security.

A potential trade-off might be a KDF not relying on CPU which is meaningless as we have seen, but using lots of RAM, such as argon2 (based on blake2b already in PLC). This might partially protect against brute-force attacks with specialized hardware (GPU, FPGA, ASIC) but the x1000 performance factor relative to a C implementation would still be there...

So the bottom line is, key derivation is not a good fit for pure Lua crypto.  Relying on such a pure Lua KDF would only give the user a false sense of security.

**Comment by catwell** [comment]
Agreed, I don't think you should use PLC in production code on things like Web servers anyway.

It can still be useful to have PBKDF2 for two reasons: 1) educational purpose and 2) use in tools where speed doesn't matter.

Regarding speed I think we can do better than 1000x slower than C. A typical value for the iterations count of PBKDF2 is 10000 so the goal would be to do that many iterations in a reasonable time. The speed of PBKDF2 depends on the hash function used so the factor should be approximately the same as SHA2, for instance.

Also, in practice yes, you should switch to Argon2 for new systems (I had one that used PBKDF2 for a long time which I switched to Argon2 recently).

It would still be interesting to see how slow a pure Lua implementation would be, both with PUC Lua and LuaJIT.

**Comment by prographo** [comment]
Ah, yes, speed. You are probably right.

I was thinking to implement a cryptocurrency wallet (BIP39/44) using pure Lua for mobile phones. One step would be calling PBKDF2 using sha512 for 2048 iterations. That's going to be really really slow, isn't it. :)

Probably I have to use C bindings.

**Comment by catwell** [comment]
> One step would be calling PBKDF2 using sha512 for 2048 iterations. That's going to be really really slow, isn't it.

Like I said, I don't think it would be *that* slow if you do it only once to obtain a key from a passphrase.

However 2048 iterations is really low. The latest guidelines from almost every organization (including NIST) recommend at least 10000.

**Comment by prographo** [comment]
I guess it's hard to predict how slow it will be - anything under 10 seconds would be fine - because as you say, this is a one time thing.

The "key" generated is a "Seed" which is then used to generate a MasterPrivateKey which is further used to generate wallet addresses (public keys). 

The Seed generation uses the PBKDF2; Inputs are a Mnemonic (a string: which is variable length entropy presented as a series of words in a particular language) + a Salt (a string: "mnemonic"+password) and outputs -> a 512bit Seed. 

The BIP39 standard specifies 2048 iterations, so it will always be that, you can read it here: https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki

How long do you think a pure Lua implementation on a mobile phone would take to execute?

My plan was to create a pure Lua wallet for a cryptocurrency and to open source this. There are other cryptographic steps apart from Seed generation - I should probably carefully look through the entire BIP39, BIP32, BIP43 & BIP44 specifications to see what other crypto primitives are required.

**Comment by philanc** [comment]
> [catwell:] Regarding speed I think we can do better than 1000x slower than C. A typical value for the iterations count of PBKDF2 is 10,000 to 100,000 so the goal would be to do that many iterations in a reasonable time. The speed of PBKDF2 depends on the hash function used so the factor should be approximately the same as SHA2, for instance.  (...)  It would still be interesting to see how slow a pure Lua implementation would be, both with PUC Lua and LuaJIT.

Agreed, we need to test!  My "1000x slower than C" was pulled out of (almost) nowhere :-)

Regarding LuaJit, I think it is out. Last time I checked it didn't support 64-bit integers and bit operators. But maybe Mike Pall changed his position on that.

...Just did a quick test  --and Ouch!! found a bug in sha2.sha512 padding :-( -- will fix it ASAP.

On my average/old laptop (i5, 2.27 MHz, linux x86_64)
    10,000 iteration of sha2.sha256: 0.58 sec
    10,000 iteration of sha2.sha512: 0.80 sec

**Comment by philanc** [comment]
> How long do you think a pure Lua implementation on a mobile phone would take to execute?

No (serious) idea. On recent phones with aarch64 CPUs, may 2x slower than on an average PC. On older phones (32-bit ARMv7), maybe 5-10x slower?

 Pierre, do you have some experience here?

> My plan was to create a pure Lua wallet for a cryptocurrency and to open source this. There are other cryptographic steps apart from Seed generation - I should probably carefully look through the entire BIP39, BIP32, BIP43 & BIP44 specifications to see what other crypto primitives are required.

I think you would need SHA2-256, SHA2-512, a HMAC construct for SHA512 (not difficult -- maybe this could be a nice addition here),  but also crypto that is not available today in PLC: RIPEMD160 (for generating key identifiers, as bitcoin addresses), and the biggest missing chunk: secp256k1 elliptic curve crypto.

**Comment by prographo** [comment]
It's not actually Bitcoin, I was thinking Cardano. I'm looking at their reference implementation now, what I have determined so far:

Seed to MasterPrivateKey: SHA512, 
Derive child PrivateKey from PrivateKey (hard): HMAC-SHA512, ed25519,
Derive child PrivateKey from PrivateKey (soft): HMAC-SHA512,
Derive child PublicKey from PublicKey (must be soft): HMAC-SHA512,
Get Public key from it's private key: ed25519
Address from public key: cbor encoding, base58
Signing transactions: ed25519::signature_extended
General bits: curve maths - curve25519 sc_reduce, ge_scalarmult_base

I think that's it. This is the core file that contains the cryptography: https://github.com/input-output-hk/rust-cardano/blob/master/cardano/src/hdwallet.rs and here is a bit of explanation: https://help.cardano.org/english/why-cardano/signatures

It's complicated, to be honest now I've looked at their code, I think it's beyond my skill level.

**Comment by philanc** [comment]
Ok. will look at it out of curiosity. Thanks for the links.

What do you mean by "hard / soft / must be soft" regarding the derivation functions?

The base curve25519 operations (scalar multiplication) are implemented in PLC. Alas, ed25519 signature is not (yet?) implemented. I am not sure if or when I'll do it.

**Comment by prographo** [comment]
The timings are fast, I think it would be quick enough.

If you do decide to add all the bits, let me know - I might give it a shot - I have the reference implementation, so if I just test every single function, really really well against their code, then I will know if it is all correct. My great fear you see, is releasing something that fails in such a way that people lose funds. That's the thing with making a wallet, stakes are kind of high.

Hard/Soft: 

This is a bit of a weird thing and I am not really convinced of the value, however it is now the standard.

The key derivation function is used to create a child key, then that child key can be used to create it's child, and so on to any level beneath. The essential inputs to the KDF are a key and a chain code. These are kept together (simple concatenation: 256bit key + 256bit chain code) and referred to as an extended key. 

An extended private key (xpriv) can be used to derive child extended private keys. An extended public key (xpub) can be used to derive child extended public keys only. Because a public key can be derived from a private key, the xpriv can also derive the corresponding xpub and all descendant xpubs.

The "normal" derivation (soft) is to use the public key as the input to the KDF for both public and private key derivation, the parent private or public key is recombined with the output to make the child.

The ability to derive a branch of xpubs from an xpub is very useful, but comes with a potential risk. Access to the xpub does not give access to child private keys - however, because the xpub contains the chain code, if a child private key becomes known, it will be possible (with the public known chain code) to derive all child public keys. And if a parent chain code is known, the parent private key can be deduced.

So.......

To counter this possibility, for private key derivation only, the BIP44 standard makes available an alternative derivation function called hardened derivation, which "breaks" the relationship between parent public key and child chain code. The parent public key is not used as an input into the KDF, the parent private key is used instead.

A Hierarchical Deterministic wallet has a number of levels, for the higher levels hardened derivation is used.

It does not change any cryptographic primitives, only how they are used; the arguments supplied.

Did that make sense? Probably not much. This video may help. https://www.youtube.com/embed/2HrMlVr1QX8?start=275

**Comment by philanc** [comment]
Thanks for the hard/soft explanation.  

So for Cardano, the big missing block in PLC is ed25519 signature. If I implement it, I will let you know.

**Comment by catwell** [comment]
> > How long do you think a pure Lua implementation on a mobile phone would take to execute?
> 
> No (serious) idea. On recent phones with aarch64 CPUs, may 2x slower than on an average PC. On older phones (32-bit ARMv7), maybe 5-10x slower?
> 
> Pierre, do you have some experience here?

I do but yeah, it varies wildly between phones. The only sure way to know is to implement and benchmark on the phones you intend to support.

If you want to implement something as complex as Cardano for a real application though, I would suggest using libsodium (in C), with bindings for the functions you need to call from Lua. I do have experience doing that, I do it at work. It works well on both Android and iOS.
